bindings/python/iree/runtime/array_interop.py

```diff
@@ -0,0 +1,232 @@
+      raise ValueError(
+          "DeviceArray cannot be implicitly transferred to the host: "
+          "if necessary, do an explicit transfer via .to_host()")
+    self._mapped_memory, self._host_array = self._map_to_host()
+
+  def _map_to_host(self) -> Tuple[MappedMemory, np.ndarray]:
+    # TODO: When synchronization is enabled, need to block here.
+    raw_dtype = self._get_raw_dtype()
+    mapped_memory = self._buffer_view.map()
+    host_array = mapped_memory.asarray(self._buffer_view.shape, raw_dtype)
+    # Detect if we need to force an explicit conversion. This happens when
+    # we were requested to pretend that the array is in a specific dtype,
+    # even if that is not representable on the device. You guessed it:
+    # this is to support bools.
+    if self._override_dtype is not None and self._override_dtype != raw_dtype:
+      host_array = host_array.astype(self._override_dtype)
+    return mapped_memory, host_array
+
+  def _get_raw_dtype(self):
+    return HalElementType.map_to_dtype(self._buffer_view.element_type)
+
+  @property
+  def dtype(self):
+    if self._override_dtype:
+      return self._override_dtype
+    return self._get_raw_dtype()
+
+  @property
+  def shape(self):
+    return np.shape(self)
+
+  def astype(self, dtype, casting="unsafe", copy=True):
+    if self.dtype == dtype and not copy:
+      return self
+    host_ary = self.to_host()
+    return host_ary.astype(dtype, casting=casting, copy=copy)
+
+  def __reduce__(self):
+    # Since this is used for making deep copies and pickling, we map
+    # separately from any interactive state. We just reduce to the actual
+    # host ndarray, which supports the necessary serialization protocols.
+    _, host_array = self._map_to_host()
+    return _restore_reduced_array, (host_array,)
+
+
+def _restore_reduced_array(ary):
+  return ary
+
+
+# Function implementations with custom behavior.
+@_device_implements(np.shape)
+def _(arr: DeviceArray):
+  return arr._buffer_view.shape
+
+
+def asdevicearray(device: HalDevice,
+                  a,
+                  dtype=None,
+                  *,
+                  implicit_host_transfer: bool = False,
+                  memory_type=MemoryType.DEVICE_LOCAL |
+                  MemoryType.DEVICE_VISIBLE,
+                  allowed_usage=BufferUsage.ALL,
+                  element_type: Optional[HalElementType] = None) -> DeviceArray:
+  """Helper to create a DeviceArray from an arbitrary array like.
+
+  This is similar in purpose and usage to np.asarray, except that it takes
+  a device as the first argument. This may not be the best mechanism for
+  getting a DeviceArray, depending on your use case, but it is reliable
+  and simple.
+
+  Note that additional flags `memory_type`, `allowed_usage` and `element_type`
+  are only hints if creating a new DeviceArray. If `a` is already a DeviceArray,
+  they are ignored.
+  """
+  if isinstance(a, DeviceArray):
+    if dtype is None:
+      return a
```

Comment:
one thing that may be tricky here for users is that this method sometimes returns a copy and sometimes returns the same instance - may be worth noting that more explicitly in the docstring

Reply:
Welcome to numpy. Copies are often made defensively in ambiguous cases. Agreed it isn't great. This function will get a lot of ergonomic work in the future that will impact when copies/transfers have to happen.

Added:
```
This function may make a defensive copy or cause implicit
  transfers to satisfy the request. If this is important to you, then a lower
  level API is likely more appropriate.
```